### PR TITLE
[dev] moe(feat): support chunkwise context parallelism for GDN.

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
# What does this PR do ?

PR for main: https://github.com/NVIDIA/Megatron-LM/pull/6439

FLA gives another CP implementation for GDN https://github.com/fla-org/flash-linear-attention/pull/691, which is claimed to outperform the old A2A-style CP. Integrate the new one into MCore.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 1]
```diff
@@ -383,29 +448,34 @@ def test_parallel_gated_delta_net_correctness(tmp_path_dist_ckpt, sequence_packi
...
         atol=atol,
         rtol=rtol,
+        cosine_similarity_threshold=cosine_similarity_threshold,
         tp=tp,
         sp=sp,
         cp=cp,
         seed=123,
         sequence_length=256,
-        micro_batch_size=4,
+        micro_batch_size=1 if (cp_comm_type == "all_gather" and cp > 1) else 4,
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply tests/unit_tests/ssm/test_gated_delta_net.py]: Disable the path `if cp_size_all_gather > 1 and packed_seq_params is None and batch > 1`.

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/6315e1578d1bfb26b625b01fc3e72e50b6c5deb2

File: megatron/core/context_parallel_layout.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    pos_in_seq = positions - global_starts[seq_idx]
+    chunk_lens = (seq_lens // chunk_divisor)[seq_idx]
+    chunk = pos_in_seq // chunk_lens
+    offset = pos_in_seq - chunk * chunk_lens
+
+    if layout == "zigzag":
+        owner = torch.where(chunk < cp_size, chunk, 2 * cp_size - chunk - 1)
+        local_slot = torch.where(chunk < cp_size, torch.zeros_like(chunk), torch.ones_like(chunk))
+    else:
+        owner = chunk // 2
```
Comment (Victarry) [inline megatron/core/context_parallel_layout.py]: **[CRITICAL Correctness]** Make the THD target layout match FLA's flattened rank-contiguous partition.

The root cause is that two different meanings of “contiguous CP layout” are being conflated. For SBHD with one logical sequence, assigning rank `r` chunks `[2r, 2r+1]` is equivalent to assigning it one contiguous interval of the global token stream. This implementation extends that rule independently to every packed sequence. FLA does not use that per-sequence partition: `build_cp_context()` computes `part_len = total_tokens // world_size` from the entire flattened THD buffer and assumes rank `r` physically owns `[r * part_len, (r + 1) * part_len)`. These layouts agree for a single sequence, which is why the current chunkwise packed test with `micro_batch_size=1` does not expose the bug, but they diverge as soon as the pack contains multiple sequences.

For example, with `cu=[0,16,40]` and `CP=2`, this code assigns rank 0 `[0:8] + [16:28]`, so its local rows contain 8 tokens from sequence 0 followed by 12 from sequence 1. FLA instead assumes rank 0 owns flat range `[0:20]` and derives `local_cu_seqlens=[0,16,20]`, interpreting the first 16 local rows as sequence 0. Consequently both causal conv and GDR consume incorrect local sequence boundaries. The state communication is also structurally incompatible: FLA only carries state for the first/last local sequence that crosses a flat rank boundary, whereas this per-sequence layout makes every packed sequence cross every rank.

**Suggested change:** keep the existing per-sequence mapping only for `layout == "zigzag"`, but define the THD `"contiguous"` indices directly from FLA's flat partition, e.g.:

```python
part_len = total_tokens // cp_size
rank_start = cp_rank * part_len
return positions[rank_start : rank_start + part_len]
```

This makes both `zigzag -> contiguous` and its inverse compatible with `build_cp_context()` without changing the A2A routing structure. Please update the current expected indices (`rank0 == range(0, 20)`, `rank1 == range(20, 40)` for this example) and add a distributed round-trip/GDN parity test with at least two unequal-length packed sequences; a single-sequence test cannot distinguish the two layouts.

File: megatron/core/ssm/gated_delta_net.py [hunk 3]
```diff
@@ -45,6 +49,7 @@
 try:
     from fla.modules.convolution import causal_conv1d
     from fla.modules.l2norm import l2norm
+    from fla.ops.cp import build_cp_context
```
Comment (Victarry) [inline megatron/core/ssm/gated_delta_net.py:52]: **[IMPORTANT Compatibility]** Raise the FLA dependency floor for `fla.ops.cp`. The current `flash-linear-attention~=0.4.0` constraint permits 0.4.0 and 0.4.1, but neither version contains `fla.ops.cp`; the API used here appears in 0.4.2. Because this import shares the top-level `try`, those supported versions set `HAVE_FLA=False` and disable even non-chunkwise GDN. Please require `flash-linear-attention>=0.4.2,<0.5` and update the lock metadata.

File: megatron/core/transformer/transformer_config.py [hunk 4]
```diff
@@ -1500,6 +1525,18 @@ def __post_init__(self):
                 f"{self.linear_num_value_heads=} must be a multiple of "
                 f"({self.tensor_model_parallel_size=} * {self.context_parallel_size=})."
             )
+            if self.context_parallel_size > 1:
+                assert self.linear_cp_mode in ("headwise", "chunkwise"), (
```
Comment (Victarry) [inline megatron/core/transformer/transformer_config.py:1519]: **[IMPORTANT Correctness]** Do not apply the headwise `TP * CP` divisibility requirement to chunkwise mode. The unconditional assertions immediately above this block still require both head counts to be divisible by `tensor_model_parallel_size * context_parallel_size`. That contradicts the new GDN initialization logic, where chunkwise CP keeps all TP-local heads and intentionally skips CP head splitting. Please require divisibility by TP only for chunkwise mode, and by `TP * CP` for headwise mode.

File: megatron/core/context_parallel_layout.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    if cp_size == 1:
+        return x
+    cp_rank = cp_group.rank()
+
+    if seq_dim != 0:
+        x = x.movedim(seq_dim, 0)
+    x = x.contiguous()
+
+    cu = cu_seqlens.to(device=x.device, dtype=torch.long)
+    source_by_rank = [
```
Comment (Victarry) [inline megatron/core/context_parallel_layout.py:145]: **[IMPORTANT Performance]** Precompute and reuse THD routing metadata instead of applying `lru_cache` directly to the Tensor-keyed helper. Each swap calls this function for both layouts and every CP rank, so one GDN layer rebuilds `arange`/`bucketize`/`argsort` metadata roughly `4 * CP` times across its two conversions, with additional host synchronizations from `.item()` and Tensor-to-bool checks. A normal `lru_cache` is unsafe because Tensor keys are identity-based and may be mutated in static CUDA-graph buffers, and the cache retains CUDA tensors. Please build a scoped/versioned routing context once per immutable packed-sequence metadata and reuse it for both directions and across GDN layers.

File: megatron/core/ssm/gated_delta_net.py [hunk 6]
```diff
@@ -417,14 +509,35 @@ def _forward_compute(
...
 
+        # Chunkwise CP expects the contiguous-time chunk layout (rank r holds chunks
+        # [2r, 2r+1]) inside conv1d / chunk_gated_delta_rule. Megatron attention CP
+        # feeds us the zigzag attention-load-balanced layout (rank r holds
+        # [r, 2*cp-r-1]), so reshuffle chunks over the CP group with a single
+        # all-to-all — no full-sequence gather required.
+        if cp_size_chunkwise > 1:
+            nvtx_range_push(suffix="zigzag_to_contiguous")
+            if packed_seq_params is not None and packed_seq_params.qkv_format == 'thd':
+                qkvzba = zigzag_to_contiguous_chunks(
```
Comment (Victarry) [inline megatron/core/ssm/gated_delta_net.py:522]: **[SUGGESTION Performance]** Consider moving the chunkwise layout conversions outside the input and output projections. Sequence-layout permutation commutes with tokenwise linear projections. For the TP=1 Qwen3.5 configurations, the current two A2As operate on summed per-token widths of 16,448 elements for 35B and 28,800 for 397B, whereas converting `hidden_states` before `in_proj` and after `out_proj` would operate on `2 * hidden_size`: 4,096 and 8,192 elements respectively. This reduces payload by approximately 4.02x and 3.52x. The implementation still needs to account for TP sequence-parallel all-gather/reduce-scatter changing the sequence distribution around these projections.

File: megatron/core/ssm/gated_delta_net.py [hunk 7]
```diff
@@ -463,11 +598,26 @@ def _forward_compute(
...
+        norm_out = norm_out.reshape(batch, seq_len_post_headwise, -1)
         norm_out = norm_out.transpose(0, 1).contiguous()
 
+        # Inverse of the zigzag -> contiguous reshuffle performed before conv1d.
+        # Restores the Megatron attention-load-balanced layout that downstream
+        # layers and loss computation expect.
+        if cp_size_chunkwise > 1:
+            nvtx_range_push(suffix="contiguous_to_zigzag")
+            if packed_seq_params is not None and packed_seq_params.qkv_format == 'thd':
+                norm_out = contiguous_to_zigzag_chunks(
```
Comment (Victarry) [inline megatron/core/ssm/gated_delta_net.py:614]: **[IMPORTANT Performance]** Avoid restoring and re-entering the layout for every GDN in a consecutive linear-attention region. With `linear_attention_freq=4`, Qwen3.5 uses three consecutive GDN layers followed by SDPA, so the current module-local contract performs six layout A2As per region: one entry and one exit for each GDN. Normalization, residual/BDA, and dense MLP operations are tokenwise, so the block could enter the contiguous layout once before the first GDN and restore zigzag only before the following SDPA or another canonical-layout boundary. Simply removing this inverse is not sufficient because the residual currently remains zigzag; layout ownership and relevant per-token metadata need to be handled at the transformer-block/region boundary.

File: tests/unit_tests/transformer/test_attention.py [hunk 8]
```diff
@@ -862,27 +863,37 @@ def get_tensor_on_this_rank(tensor):
...
+                torch.testing.assert_close(
+                    baseline,
+                    parallel,
+                    atol=atol,
+                    rtol=rtol,
+                    msg=lambda msg: f"Mismatch in {tensor_name}: {msg}",
+                )
+                return
+            except AssertionError as close_error:
+                if cosine_similarity_threshold is None:
```
Comment (Victarry) [inline tests/unit_tests/transformer/test_attention.py:878]: **[IMPORTANT Numerical Testing]** Do not use cosine similarity as the sole fallback after `assert_close` fails. Cosine similarity is insensitive to scale, so even `parallel = 100 * baseline` passes with a score of 1.0. This can hide material output and gradient errors in the new CP path. Please require an additional relative-L2/norm-ratio bound, or use explicitly relaxed elementwise tolerances together with cosine similarity.

File: megatron/core/context_parallel_layout.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    if cu_seqlens.dim() != 1:
+        raise ValueError(f"cu_seqlens must be 1-D, got shape {tuple(cu_seqlens.shape)}.")
+
+    cu = cu_seqlens.to(dtype=torch.long)
+    if cu.numel() == 0 or cu[0].item() != 0:
+        raise ValueError(f"cu_seqlens must start at 0, got {cu_seqlens}.")
+
+    seq_lens = torch.diff(cu)
+    chunk_divisor = 2 * cp_size
+    if torch.any(seq_lens <= 0):
```
Comment (xiaoyao0115) [inline megatron/core/context_parallel_layout.py]: **[IMPORTANT Compatibility] Allow zero-length placeholder entries in statically padded THD `cu_seqlens`.**

In static THD CUDA Graph mode, Megatron pads `cu_seqlens` to `thd_max_packed_sequences + 1` entries by repeating the final offset. For example, a valid `[0, 16, 40]` may become `[0, 16, 40, 40, 40]`. The repeated trailing entries are placeholders used to keep the metadata shape static, and therefore produce zero-length segments.

The current `seq_lens <= 0` check rejects this valid representation, causing chunkwise CP to fail whenever the actual number of packed sequences is smaller than the configured maximum.

Please reject only decreasing boundaries, and ignore zero-length placeholder segments when validating divisibility and computing ownership, or otherwise normalize repeated boundaries without changing the static input contract. Please also add a test with trailing duplicate offsets.

File: megatron/core/ssm/gated_delta_net.py (hunk 6 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py:522]: Very good suggestion, but we plan to make this change in the upcoming refactor works. Will add a comment for this.

File: megatron/core/ssm/gated_delta_net.py (hunk 7 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py:614]: Similar to https://github.com/NVIDIA/Megatron-LM/pull/3282#discussion_r3490324819, I will leave a comment to explain.

File: megatron/core/context_parallel_layout.py (hunk 5 above)
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout.py:145]: Good point. The routing is a function of cu_seqlens, so it is only reusable within the same packed microbatch. Since cu_seqlens changes across batches, a persistent cache would be unsafe unless keyed by the actual sequence-boundary contents, which is likely not worth the synchronization/complexity. I added a TODO to keep this as future work for a per-microbatch CP layout context instead of caching it here.

File: tests/unit_tests/transformer/test_attention.py (hunk 8 above)
Comment (yuzhongw-nvidia) [reply tests/unit_tests/transformer/test_attention.py:878]: Very good point. Resolved.

File: megatron/core/context_parallel_layout.py (hunk 9 above)
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout.py]: Thanks. Resolved.

File: megatron/core/transformer/transformer_config.py (hunk 4 above)
Comment (yuzhongw-nvidia) [reply megatron/core/transformer/transformer_config.py:1519]: Very good catch. Resolved.

File: megatron/core/ssm/gated_delta_net.py (hunk 3 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py:52]: Thanks. Resolved.

File: megatron/core/context_parallel_layout.py (hunk 2 above)
Comment (yuzhongw-nvidia) [reply megatron/core/context_parallel_layout.py]: Very good catch. Helps a lot! Resolved.

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/644098c7228b2c72187108f3a2a20394bf19df26

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3282/commits/df71688f58c82eeba6d6dbcdd723360b69a09282

**Review by Victarry — APPROVED** [review]
LGTM!Thanks for the great work~

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/6aaf4b78d6938feb835cc91f954c533a9f6cd2aa

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/0beec47dd0ebf7135be07ee1ece8157fb8eae308

**Review by xiaoyao0115 — APPROVED** [review]
lgtm！

**Review by ko3n1g — APPROVED** [review]
